Implement comprehensive documentation with MkDocs and arc42

groldan opened the pull request:
## Summary
- Migrated existing AsciiDoc documentation to MkDocs with Material theme for improved organization and readability
- Added comprehensive architecture documentation following the arc42 template, including system context, container, component views, and more
- Created migration guide from security-proxy to Gateway with configuration mappings and examples
- Added clear distinction between development and production configurations and environments

## Live Demo
A live version of the documentation is available at: https://groldan.github.io/georchestra-gateway/

## Details

### Documentation Structure
- Created user guide, developer guide, and technical documentation sections
- User guide covers installation, configuration, authentication, and other operational topics
- Developer guide includes building, testing, contributing, and implementation details
- Technical documentation follows arc42 template with architecture views and diagrams

### Key Documentation Components
- **Migration Guide**: Comprehensive path from security-proxy to Gateway
- **Architecture Documentation**: Complete arc42 documentation with diagrams
- **Configuration Documentation**: Detailed explanations of all configuration options
- **Development Setup**: Clear instructions for developers with warnings about dev-only configs

### Technical Implementation
- Added support for MkDocs with Material theme
- Implemented Structurizr DSL for architecture diagrams
- Added scripts for generation of SVG diagrams
- Updated GitHub workflow for documentation generation
- Added README files to clarify development vs production resources

### Development Environment Clarification
- Added DOCKER_COMPOSE_README.md to explain development-only Docker compose files
- Updated datadir README to clarify it's for development use only
- Added warnings in user guide installation page about not using dev configs in production

## Testing
- Verified diagram generation works with the provided scripts
- Confirmed MkDocs site builds correctly with `mkdocs build`
- Checked all links and navigation in the documentation
- Ensured clear warnings about development vs production usage

These changes significantly improve the documentation quality and organization, providing clear guidance for both users and developers of the Gateway component.

---

Fixes #154

landryb commented:
https://groldan.github.io/georchestra-gateway/arc42/runtime_view/#deployment-considerations

> The Gateway is designed to be deployed as a Docker container, with all configuration externalized in a data directory volume.

but can also be deployed as a standalone spring-boot service outside docker environments.

landryb commented:
that's a cosmetic thing, but many (all?) lists of items are rendered on a single line:

> Key components: - ReactorContextHolder: Ensures MDC context is maintained in reactive code. - AccessLogFilter: Records HTTP requests with pattern-based filtering. - MdcLogEnhancerFilter: Adds context information to logs.

landryb commented:
another personal feeling, but tons of content feel like AI-generated, which might be fine, but .. feels awkward. Especially with the moniker *Made by people for people* at the bottom :)

jeanpommier commented on docs/developer_guide/building.md:
````diff
@@ -0,0 +1,201 @@
+./mvnw formatter:validate -pl :georchestra-gateway
+```
+
+## Development Environment
````
Maybe some info are missing. Something I usually like is also a simple _Run it in the commandline_ instruction 
(`./mvnw spring-boot:run -Dspring-boot.run.profiles=json-logs,dev -pl :georchestra-gateway`)
which also shows how to target specific profiles. 
We can also provide the `SPRING_PROFILES_ACTIVE=json-logs,dev` way)

groldan replied:
* [fix](https://github.com/georchestra/georchestra-gateway/compare/790b7da38a57ecbbade37684744ca04cc604fce0..71b81fb7371059f9329481d46d8e0d9bda2157d9#diff-9ead6f7a795a284273759f052cb49ffcdd3aba154260dff5dcc756fb671ed766R133)
* live: https://groldan.github.io/georchestra-gateway/developer_guide/building/#running-from-the-command-line

jeanpommier commented on docs/user_guide/index.md:
```diff
@@ -0,0 +1,42 @@
+
+geOrchestra Gateway relies on the geOrchestra "data directory" for its configuration. The relevant files are:
+
+- `default.properties` - Common configuration properties
```
not very clear. "Configuration properties common to all geOrchestra applications"

groldan replied:
it is what it is

jeanpommier commented on docs/user_guide/installation.md:
````diff
@@ -0,0 +1,293 @@
+    - "5005:5005"  # Expose debug port
+```
+
+### Using Kubernetes
````
Kubernetes in the docker section ? I'd rather have it in a separate section. And would mention the geOrchestra Helm chart

groldan replied:
yeah, doesn't matter that's it's docker, they're two completely separate deployment models. Check out [these changes](https://github.com/georchestra/georchestra-gateway/compare/d1b355eda77aaf88691f5fa1a2f49c8a3a6323b0..d096bea6af1d0cb0c2279e213e0f3bcb0463ab87) or https://groldan.github.io/georchestra-gateway/user_guide/installation/#installation-options

jeanpommier commented on docs/user_guide/installation.md:
```diff
@@ -0,0 +1,293 @@
+
+After installation, configure the Gateway by editing the files in the geOrchestra data directory:
+
+1. `default.properties` - Common configuration properties
```
not very clear. "Configuration properties common to all geOrchestra applications"

groldan replied:
isn't that what default.properties is in the georchestra datadir? how would you phrase it otherwise?

jeanpommier replied:
"Common configuration properties" is not clear about to whom this is common. That's why I'd rephrase to "Configuration properties common to all geOrchestra applications"

jeanpommier commented on docs/user_guide/configuration.md:
````diff
@@ -0,0 +1,399 @@
+```
+
+Common profiles include:
+- `docker` - Settings for Docker environments
````
Missing linebreak, the list doesn't render as a list

jeanpommier commented on docs/user_guide/configuration.md:
```diff
@@ -0,0 +1,399 @@
+- `docker` - Settings for Docker environments
+- `json-logs` - Enable JSON-formatted structured logging
+- `logging_debug` - Enable more detailed logging
+- `preauth` - Enable pre-authentication mode
```
Not self-explanatory. Would be good to provide an href there (https://groldan.github.io/georchestra-gateway/user_guide/authentication/#pre-authentication-via-http-headers right ?)

groldan replied:
[gotcha](https://github.com/georchestra/georchestra-gateway/compare/d096bea6af1d0cb0c2279e213e0f3bcb0463ab87..521fa3e62c58dcabc0fbc567fa0315f68677a7b8)

jeanpommier commented on docs/user_guide/ui_customization.md:
```diff
@@ -0,0 +1,170 @@
+1. Creating your own version of the header component
+2. Updating the `headerScript` property to point to your custom implementation
+3. Or completely disabling the header by setting `georchestra.gateway.headerEnabled: false`
+
```
Maybe mention that you can now [configure the header's menu entries](https://github.com/georchestra/header/blob/main/CONFIG.md)

groldan replied:
Would that be part of the gateway configuration or the header application?

jeanpommier replied:
That's configuring the header, in the gateway application. It's mostly heaer-related, but worth mentioning here. Probably just mentioning it with a href to the header doc would be enough, I'd say

jeanpommier commented on docs/user_guide/ui_customization.md:
````diff
@@ -0,0 +1,170 @@
+          filters:
+            - name: ApplicationError
+```
+
````
Be more specific as to what it does. Not clear

jeanpommier commented on docs/developer_guide/custom_filters.md:
```diff
@@ -0,0 +1,341 @@
+**Class**: `org.georchestra.gateway.filter.headers.AddSecHeadersGatewayFilterFactory`
+
+**Header Contributors**:
+- `GeorchestraUserHeadersContributor`: Adds user-related headers (username, email, roles)
```
missing linefeed

jeanpommier commented on docs/developer_guide/custom_filters.md:
````diff
@@ -0,0 +1,341 @@
+      default-filters:
+        - ApplicationError
+```
+
````
please explain how the custom error response is provided (custom error template ?)

jeanpommier commented on docs/developer_guide/authentication.md:
```diff
@@ -0,0 +1,175 @@
+- `OpenIdConnectUserMapper`: Maps OAuth2 authentication to GeorchestraUser objects
+- `OpenIdConnectCustomConfig`: Customization for specific providers
+
+To customize OAuth2 authentication:
```
Missing linefeed

jeanpommier commented on docs/developer_guide/authentication.md:
```diff
@@ -0,0 +1,175 @@
+- `PreauthAuthenticationManager`: Validates pre-authentication headers
+- `PreauthenticatedUserMapperExtension`: Maps header values to GeorchestraUser objects
+
+To customize header pre-authentication:
```
Missing linefeed

jeanpommier commented on docs/developer_guide/authentication.md:
```diff
@@ -0,0 +1,175 @@
+- `RolesMappingsUserCustomizer`: Applies role mappings based on configuration
+- `CreateAccountUserCustomizer`: Handles user creation through login
+
+To implement a custom user customizer:
```
Missing linefeed

jeanpommier commented on docs/arc42/architecture_goals.md:
```diff
@@ -0,0 +1,118 @@
+**Rationale**: As the entry point for all user interactions, the Gateway must enforce robust security controls to protect sensitive geospatial data and services.
+
+**Approaches**:
+- Implement multiple authentication methods (LDAP, OAuth2/OpenID Connect, Header pre-authentication)
```
missing linefeed. Idem on later `Approaches` sections

jeanpommier commented on docs/arc42/context_view.md:
```diff
@@ -0,0 +1,142 @@
+
+1. **Users** access the geOrchestra platform through the Gateway using HTTPS
+2. **External Identity Providers** can authenticate users via OAuth2/OpenID Connect
+3. **geOrchestra Gateway** acts as the entry point for all services
```
not really shown actually

groldan replied:
I've updated the context_view.md document to correctly reflect what's actually shown in the system context diagram.
* [Fix](https://github.com/georchestra/georchestra-gateway/compare/b0de5620faab254021e15451c9beb441fb941d77..238db3b807f84932d4130ff5ff556dde2be8bfaf)
* [Live](https://groldan.github.io/georchestra-gateway/arc42/context_view/#system-context-diagram)

jeanpommier replied:
wouldn't it rather be the gateway intead of thegeOrchestra SDI ?

jeanpommier commented on docs/arc42/context_view.md:
```diff
@@ -0,0 +1,142 @@
+4. **geOrchestra SDI Platform** contains the various backend services
+5. The Gateway interacts with external systems:
+   - **LDAP Directory** for user authentication and attributes
+   - **Database** (PostgreSQL) for session storage
```
?

jeanpommier commented on docs/arc42/context_view.md:
```diff
@@ -0,0 +1,142 @@
+- **Users** interact with the Gateway via HTTPS requests
+- **External Identity Providers** can authenticate users via OAuth2/OpenID Connect
+- The **Gateway** routes authenticated requests to the various geOrchestra services
+- External systems like **LDAP**, **Database**, and **Message Broker** support the Gateway's authentication, storage, and event notification needs
```
Not the database

jeanpommier commented on docs/arc42/context_view.md:
```diff
@@ -0,0 +1,142 @@
+- HTTP/HTTPS for web traffic
+- LDAP for directory service communication
+- AMQP for message broker communication
+- SQL for database communication
```
nope, don't think so

jeanpommier commented on docs/arc42/container_view.md:
```diff
@@ -0,0 +1,78 @@
+### Target Backend Services
+
+The Gateway routes requests to various geOrchestra platform services:
+- **GeoServer**: OGC-compliant map and feature server
```
missing linefeed

jeanpommier commented on docs/arc42/deployment_view.md:
```diff
@@ -0,0 +1,283 @@
+
+## Overview
+
+The geOrchestra Gateway is designed to be deployed as a containerized application, typically using Docker. It is part of the broader geOrchestra Spatial Data Infrastructure (SDI) ecosystem and acts as the central entry point for all user requests to geOrchestra services.
```
Could cause friction in the community. Some people don't use docker at all and will feel excluded if we keep this statement. Since this is a spring boot app, it doesn't rely on docker. I suggest to reformulate. Maybe something like
"The geOrchestra Gateway is designed as a Spring Boot microservice, that can be run as a simple executable, a system service or be deployed as a containerized application, typically using Docker"

groldan replied:
How sensitive. How about this:

```
The geOrchestra Gateway is designed as a Spring Boot microservice that can be run
as a simple executable, a system service, or be deployed as a containerized application
(typically using Docker). It is part of the broader geOrchestra Spatial Data Infrastructure (SDI)
ecosystem and acts as the central entry point for all user requests to geOrchestra services.
```

* [Fix](https://github.com/georchestra/georchestra-gateway/compare/238db3b807f84932d4130ff5ff556dde2be8bfaf..7f3d91e2d65fe257262e34db940def6fef69bd7f)
* [Live](https://groldan.github.io/georchestra-gateway/arc42/deployment_view/#overview)
  *  [Deployment options](https://groldan.github.io/georchestra-gateway/arc42/deployment_view/#deployment-options)

jeanpommier replied:
Perfect, thanks ! ;o)

landryb replied:
a bit late to the party, but i second that version, thanks !

jeanpommier commented on docs/arc42/deployment_view.md:
```diff
@@ -0,0 +1,283 @@
+- **Ports**: Exposes port 8080 internally
+- **Dependencies**: 
+  - LDAP for authentication
+  - PostgreSQL for session storage
```
nope, not postgresql. Or is it ?

jeanpommier commented on docs/arc42/deployment_view.md:
```diff
@@ -0,0 +1,283 @@
+
+- **Path**: `/etc/georchestra` (inside the container)
+- **Contents**:
+  - `default.properties` - Common properties
```
Reformulate as "properties common to all geOrchestra applications"

jeanpommier commented on docs/arc42/deployment_view.md:
```diff
@@ -0,0 +1,283 @@
+   - Typically deployed as `georchestra/ldap` container
+   - Requires persistent volume for data
+
+2. **PostgreSQL Database**:
```
really ?

jeanpommier commented on docs/arc42/crosscutting.md:
```diff
@@ -0,0 +1,199 @@
+- **Header Pre-authentication**: For integration with upstream authentication proxies.
+
+Key classes involved:
+- `GatewaySecurityConfiguration`: Central configuration for all security components.
```
missing linefeed

jeanpommier commented on docs/arc42/crosscutting.md:
```diff
@@ -0,0 +1,199 @@
+- **UTF-8 Support**: Throughout the application.
+
+Key files:
+- `messages/login_*.properties`: Localized messages for the login page.
```
missing linefeed

jeanpommier commented on docs/arc42/crosscutting.md:
```diff
@@ -0,0 +1,199 @@
+- **Error Attributes**: Standardized error response format.
+
+Key components:
+- `CustomErrorAttributes`: Enhances error information for responses.
```
missing linefeed

jeanpommier commented on docs/arc42/crosscutting.md:
```diff
@@ -0,0 +1,199 @@
+- **RabbitMQ Integration**: For cross-service communication.
+
+Key components:
+- `RabbitmqEventsConfiguration`: Configures event publishing via RabbitMQ.
```
missing linefeed

jeanpommier commented on docs/arc42/quality_requirements.md:
```diff
@@ -0,0 +1,307 @@
+**Requirement**: The Gateway must support multiple authentication methods with industry-standard security.
+
+**Scenarios**:
+- **Scenario 1**: An attacker attempts to bypass authentication by manipulating HTTP headers. The system must detect and reject such attempts.
```
missing linefeed. Same issue on below sections

jeanpommier commented on docs/arc42/quality_requirements.md:
```diff
@@ -0,0 +1,307 @@
+- **Scenario 2**: A brute force attack against the login page is initiated. The system must implement rate limiting and account lockout mechanisms.
+
+**Measures**:
+- Support for OAuth2/OpenID Connect with current security practices
```
missing linefeed. Same issue on below sections

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+Risks are assessed using the following criteria:
+
+* **Impact**: The severity of consequences if the risk materializes
+  * High: Major service disruption, security breach, or data loss
```
missing indent. Don't know exactly why

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+**Priority**: High  
+
+**Mitigation Strategies**:
+- Regular security audits of OAuth2 implementation
```
missing linefeed

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+**Priority**: Medium  
+
+**Mitigation Strategies**:
+- Use parameterized LDAP queries
```
missing linefeed

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+**Priority**: High  
+
+**Mitigation Strategies**:
+- Strict validation of pre-authentication headers
```
missing linefeed

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+**Priority**: Medium  
+
+**Mitigation Strategies**:
+- Thorough code reviews focused on reactive patterns
```
missing linefeed

jeanpommier commented on docs/arc42/risks.md:
```diff
@@ -0,0 +1,310 @@
+**Priority**: Medium  
+
+**Mitigation Strategies**:
+- Connection pooling for LDAP
```
missing linefeed. Same in sections below

jeanpommier reviewed (commented):
That's a great, extensive doc, quite impressive. thanks @groldan !

groldan commented:
> another personal feeling, but tons of content feel like AI-generated, which might be fine, but .. feels awkward. Especially with the tagline _Made by people for people_ at the bottom :)

@landryb and yet it took me over 2 full working days to come up with the initial pull request. You'd need to hire a technical writer and have it working for a month otherwise.

Is code that used autocompletion less made by people? The tagline is about geOrchestra made by people for the people, not the docs. We can say the docs are made by people that knows how to use AI.

groldan commented on docs/user_guide/configuration.md:
```diff
@@ -0,0 +1,399 @@
+    gateway:
+      globalcors:
+        corsConfigurations:
+          '[/**]':
```
* [fix](https://github.com/georchestra/georchestra-gateway/compare/521fa3e62c58dcabc0fbc567fa0315f68677a7b8..790b7da38a57ecbbade37684744ca04cc604fce0)
* [live](https://groldan.github.io/georchestra-gateway/user_guide/configuration/#cors-configuration)

jeanpommier commented:
Really great work @groldan. I'm also not a great fan of AI, but I must recognize that this doc is very thorough. Nobody would have taken the time to document that much manually.
I just wonder about keeping it up-to-date since as @landryb said, it is very big (and with some redundancy, presenting several aspects of a same thing). 
But I suppose the answer will be "by asking AI to do it", right ? ;o)

groldan commented:
As said, it's a tool. So far, we've been having trouble keeping the documentation up to date with code changes, and I think that statement applies almost everywhere.
The important part is to make docs update part of the process. As part of the acceptance criteria for example.
Then do it manually, with AI, or ask your uncle to do it, doesn't matter.

groldan commented on docs/user_guide/configuration.md:
```diff
@@ -0,0 +1,399 @@
+- `Method`: Matches based on the HTTP method
+- `Header`: Matches based on a request header
+- `Query`: Matches based on a query parameter
+
```
* [Fix](https://github.com/georchestra/georchestra-gateway/compare/151bbbf7fbe44a915523487076cd003e7aae3d3c..2452166b115ae16ff1c11c0745890811dc736a5e)
* Live:
  - https://groldan.github.io/georchestra-gateway/user_guide/configuration/#available-predicates
  - https://groldan.github.io/georchestra-gateway/user_guide/configuration/#available-filters

jeanpommier commented:
> As said, it's a tool. So far, we've been having trouble keeping the documentation up to date with code changes, and I think that statement applies almost everywhere. The important part is to make docs update part of the process. As part of the acceptance criteria for example. Then do it manually, with AI, or ask your uncle to do it, doesn't matter.

Ha ha ha, no, my uncle is definitely not an option !

groldan commented on docs/user_guide/authentication.md:
````diff
@@ -0,0 +1,355 @@
+```
+
+The `roles-mapping` section allows you to map provider-specific roles to geOrchestra roles.
+
````
you're absolutely right! I made a mistake (now I'm impersonating an AI, or am I?)

This should do: 
* [Fix](https://github.com/georchestra/georchestra-gateway/compare/2452166b115ae16ff1c11c0745890811dc736a5e..024b5cac90fe21b10736c4b9e886dcd0a66fe3c3)
* [Live](http://127.0.0.1:8000/gateway/user_guide/authentication/#oauth2openid-connect-authentication)

groldan commented on docs/user_guide/access_rules.md:
````diff
@@ -0,0 +1,156 @@
+
+The configuration of security access rules follows this object model:
+
+```mermaid
````
Removed. A class diagram has nothing to do with the user documentation, and is already supplied in the [developer guide](https://groldan.github.io/georchestra-gateway/developer_guide/project_structure/#configuration-object-model).

* [Fix](https://github.com/georchestra/georchestra-gateway/compare/024b5cac90fe21b10736c4b9e886dcd0a66fe3c3..7ccc389cffff525906eada61ea23e499a2d94348#diff-0a4dd950db9053a7cc1699f087da78676cca96bf74c17a6a663e5c6a3478b77aL5)

Also made some notes about what I think is the biggest pita:
- [ResolveTargetGlobalFilter note in developer guid](https://groldan.github.io/georchestra-gateway/developer_guide/custom_filters/#global-filters):
- [Technical debt](https://groldan.github.io/georchestra-gateway/arc42/risks/#configuration-model)
- [Connecting Services to Routes note in the user guide](https://groldan.github.io/georchestra-gateway/user_guide/access_rules/#connecting-services-to-routes)
- [Matching Service Target with Route URI note in the user guide](https://groldan.github.io/georchestra-gateway/user_guide/proxying_applications/)

groldan commented on docs/user_guide/logging.md:
```diff
@@ -0,0 +1,174 @@
+  "thread": "http-nio-8080-exec-1",
+  "logger": "org.georchestra.gateway.logging.RequestLogger",
+  "message": "Received request",
+  "mdc": {
```
* [Fix](https://github.com/georchestra/georchestra-gateway/compare/7ccc389cffff525906eada61ea23e499a2d94348..b0de5620faab254021e15451c9beb441fb941d77)
* [Live](https://groldan.github.io/georchestra-gateway/user_guide/logging/#enabling-json-logging)

groldan commented:
@jeanpommier @landryb thank you so much for the in-depth review. I believe I've (well, we, Claude and I) addressed all the issues and this should be good to go. Let me know if there's any blocker issue remaining though.

jeanpommier commented on docs/arc42/context_view.md:
```diff
@@ -0,0 +1,139 @@
+
+1. **Users** interact with the geOrchestra SDI platform using HTTPS
+2. **geOrchestra SDI** is the overall system containing the Gateway and other services
+3. The **geOrchestra SDI** interacts with external systems:
```
It still feels a bit off. I'd replace "geOrchestra SDI" by Gateway, I believe it would be more meaningful. But still, I'm not sure this diagram brings much to the global understanding

jeanpommier replied:
Apart from this (and, well, not really a big deal), it's all good for me. 
It's getting late here, I'll disconnect. On my side, feel free to merge it

jeanpommier replied:
And thanks for your patience ! ;o)

groldan replied:
Nope, context view is context view. There's the [container view](https://groldan.github.io/georchestra-gateway/arc42/container_view/) where the gateway is shown right bellow that one :)

landryb commented:
(unless im wrong) now that this has been merged, the deployed version of the documentation is now available at https://www.georchestra.org/georchestra-gateway/
